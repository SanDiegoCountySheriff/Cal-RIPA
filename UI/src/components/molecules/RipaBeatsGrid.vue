<template>
  <div>
    <v-card-title>
      <v-text-field
        v-model="search"
        append-icon="mdi-magnify"
        label="Search"
        single-line
        hide-details
      ></v-text-field>
    </v-card-title>
    <v-data-table
      :loading="loading"
      :headers="headers"
      :items="beats"
      :search="search"
      sort-by="community"
    >
      <template v-slot:top>
        <v-toolbar flat>
          <v-toolbar-title class="tw-uppercase"
            >Admin: Maintain Beats</v-toolbar-title
          >
          <v-spacer></v-spacer>
          <v-dialog v-model="dialog" max-width="500px">
            <template v-slot:activator="{ on, attrs }">
              <v-btn
                color="primary"
                dark
                class="tw-mb-2"
                v-bind="attrs"
                v-on="on"
              >
                New Item
              </v-btn>
            </template>
            <v-card>
              <v-card-title>
                <span>{{ formTitle }}</span>
              </v-card-title>

              <v-card-text>
                <v-container>
                  <v-row>
                    <v-col cols="12">
                      <v-text-field
                        v-model="editedItem.community"
                        label="Community"
                      ></v-text-field>
                    </v-col>
                    <v-col cols="12">
                      <v-text-field
                        v-model="editedItem.command"
                        label="Command"
                      ></v-text-field>
                    </v-col>
                    <v-col cols="12">
                      <v-text-field
                        v-model="editedItem.commandAuditGroup"
                        label="Command Audit Group"
                      ></v-text-field>
                    </v-col>
                    <v-col cols="12">
                      <v-text-field
                        v-model="editedItem.commandAuditSize"
                        label="Command Audit Size"
                      ></v-text-field>
                    </v-col>
                  </v-row>
                </v-container>
              </v-card-text>

              <v-card-actions>
                <v-spacer></v-spacer>
                <v-btn color="blue darken-1" text @click="close">
                  Cancel
                </v-btn>
                <v-btn color="blue darken-1" text @click="save"> Save </v-btn>
              </v-card-actions>
            </v-card>
          </v-dialog>
          <v-dialog v-model="dialogDelete" max-width="500px">
            <v-card>
              <v-card-title
                >Are you sure you want to delete this beat?</v-card-title
              >
              <v-card-actions>
                <v-spacer></v-spacer>
                <v-btn color="blue darken-1" text @click="closeDelete"
                  >Cancel</v-btn
                >
                <v-btn color="blue darken-1" text @click="deleteItemConfirm"
                  >OK</v-btn
                >
                <v-spacer></v-spacer>
              </v-card-actions>
            </v-card>
          </v-dialog>
        </v-toolbar>
      </template>
      <template v-slot:item.actions="{ item }">
        <v-icon small class="tw-mr-2" @click="editItem(item)">
          mdi-pencil
        </v-icon>
        <v-icon small @click="deleteItem(item)"> mdi-delete </v-icon>
      </template>
      <template v-slot:no-data>
        <div>No Data</div>
      </template>
    </v-data-table>
  </div>
</template>

<script>
export default {
  name: 'ripa-beats-grid',

  data() {
    return {
      search: '',
      beats: [],
      dialog: false,
      dialogDelete: false,
      headers: [
        { text: 'ID', value: 'id' },
        { text: 'Community', value: 'community' },
        { text: 'Command', value: 'command' },
        { text: 'Command Audit Group', value: 'commandAuditGroup' },
        { text: 'Command Audit Size', value: 'commandAuditSize' },
        { text: 'Actions', value: 'actions', sortable: false, width: '100' },
      ],
      editedIndex: -1,
      editedItem: {
        community: '',
        command: '',
        commandAuditGroup: '',
        commandAuditSize: '',
      },
      defaultItem: {
        community: '',
        command: '',
        commandAuditGroup: '',
        commandAuditSize: '',
      },
    }
  },

  computed: {
    formTitle() {
      return this.editedIndex === -1 ? 'New Item' : 'Edit Item'
    },
  },

  methods: {
    init() {
      this.beats = this.items
    },

    editItem(item) {
      this.editedIndex = this.beats.indexOf(item)
      this.editedItem = Object.assign({}, item)
      this.dialog = true
    },

    deleteItem(item) {
      this.editedIndex = this.beats.indexOf(item)
      this.editedItem = Object.assign({}, item)
      this.dialogDelete = true
    },

    deleteItemConfirm() {
      this.beats.splice(this.editedIndex, 1)
      this.closeDelete()
    },

    close() {
      this.dialog = false
      this.$nextTick(() => {
        this.editedItem = Object.assign({}, this.defaultItem)
        this.editedIndex = -1
      })
    },

    closeDelete() {
      this.dialogDelete = false
      this.$nextTick(() => {
        this.editedItem = Object.assign({}, this.defaultItem)
        this.editedIndex = -1
      })
    },

    save() {
      if (this.editedIndex > -1) {
        Object.assign(this.beats[this.editedIndex], this.editedItem)
      } else {
        this.beats.push(this.editedItem)
      }
      this.close()
    },
  },

  watch: {
    items(val) {
      this.beats = val
    },
    dialog(val) {
      val || this.close()
    },
    dialogDelete(val) {
      val || this.closeDelete()
    },
  },

  created() {
    this.init()
  },

  props: {
    loading: {
      type: Boolean,
      default: false,
    },
    items: {
      type: Array,
      default: () => [],
    },
  },
}
</script>
