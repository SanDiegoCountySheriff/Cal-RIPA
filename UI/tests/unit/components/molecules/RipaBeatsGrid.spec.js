import RipaBeatsGrid from '@/components/molecules/RipaBeatsGrid.vue'
import { shallowMount } from '@vue/test-utils'
import Vuetify from 'vuetify'

describe('Ripa Beats Grid', () => {
  let vuetify
  let wrapper

  beforeEach(() => {
    vuetify = new Vuetify()
  })

  afterEach(() => {
    wrapper.destroy()
  })

  const factory = propsData => {
    return shallowMount(RipaBeatsGrid, {
      vuetify,
      propsData: {
        ...propsData,
        onEditBeat: jest.fn(),
        onDeleteBeat: jest.fn(),
        items: [
          {
            id: 1,
            community: 'community 1',
            command: 'command 1',
            commandAuditGroup: 'audit group 1',
            commandAuditSize: 'audit size 1',
          },
          {
            id: 2,
            community: 'community 2',
            command: 'command 2',
            commandAuditGroup: 'audit group 2',
            commandAuditSize: 'audit size 2',
          },
        ],
      },
    })
  }

  it('should match snapshot', () => {
    wrapper = factory()

    expect(wrapper.element).toMatchSnapshot()
  })

  it.todo('should compute formTitle')

  it.todo('should compute isRowKeyDisabled')

  it.todo('should compute isDuplicateKey')

  it.todo('should edit item')

  it.todo('should delete item')

  it.todo('should close')

  it.todo('should close delete')

  it.todo('should save')

  it.todo('should watch items')

  it.todo('should watch dialog')

  it.todo('should watch dialogDelete')
})
