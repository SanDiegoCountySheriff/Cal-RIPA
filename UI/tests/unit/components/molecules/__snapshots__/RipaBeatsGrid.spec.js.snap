// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Ripa Beats Grid should match snapshot 1`] = `
<div
  class="ripa-beats-grid"
>
  <v-card-title-stub>
    <v-text-field-stub
      appendicon="mdi-magnify"
      backgroundcolor=""
      clearicon="$clear"
      errorcount="1"
      errormessages=""
      hidedetails="true"
      label="Search"
      loaderheight="2"
      messages=""
      rules=""
      singleline="true"
      successmessages=""
      type="text"
      value=""
    />
  </v-card-title-stub>
   
  <v-data-table-stub
    customfilter="[Function]"
    customgroup="[Function]"
    customsort="[Function]"
    expanded=""
    expandicon="$expand"
    groupby=""
    groupdesc=""
    headers="[object Object],[object Object],[object Object],[object Object],[object Object],[object Object]"
    itemclass=""
    itemkey="id"
    items=""
    itemsperpage="10"
    loaderheight="4"
    loadingtext="$vuetify.dataIterator.loadingText"
    locale="en-US"
    mobilebreakpoint="600"
    nodatatext="$vuetify.noDataText"
    noresultstext="$vuetify.dataIterator.noResultsText"
    options="[object Object]"
    page="1"
    search=""
    selectablekey="isSelectable"
    serveritemslength="-1"
    sortby="community"
    sortdesc=""
    value=""
  />
</div>
`;
